django/db/models/sql/query.py

```diff
@@ -2108,3 +2108,435 @@ def update_join_types(self, query):
         query.promote_joins(to_promote)
         query.demote_joins(to_demote)
         return to_demote
+
+
+class ExpressionConstraintQuery:
```

Comment:
It may be possible to simplify this further. I'm also not 100% certain I've thought of every edge-case, so I may have accidentally deleted something important.

Reply:
This *is* a lot of stuff to duplicate, especially as you note below that you don't understand everything. I'm concerned that future expression refactors/features will be missed in constraints and things will asplode.

Reply:
Yeah - that's a good point. I'll take another pass and see if I can minimise the amount of duplicated code.

Reply:
I'm out of my depth here; my instinct would be to figure out what the responsibilities of Q / Query / etc. are, which bits are needed for constraints, and refactor to share these bits.

Malcolm's "Dungeon Master's guide to Django's ORM" talk may help.

Reply:
For example, `_force_pk` (duplicated below) was removed in ec50937bcbe160e658ef881021402e156beb0eaf.